src/workspace.coffee

```diff
@@ -791,36 +800,72 @@ class Workspace extends Model
   #   * `paths` An {Array} of glob patterns to search within
+  #   * `onPathsSearched` (optional) {Function}
   # * `iterator` {Function} callback on each file found
   #
-  # Returns a `Promise`.
+  # Returns a `Promise` with a `cancel()` method.
   scan: (regex, options={}, iterator) ->
     if _.isFunction(options)
       iterator = options
       options = {}
 
-    deferred = Q.defer()
-
     searchOptions =
       ignoreCase: regex.ignoreCase
-      inclusions: options.paths
+      inclusions: options.paths or []
       includeHidden: true
       excludeVcsIgnores: atom.config.get('core.excludeVcsIgnoredPaths')
       exclusions: atom.config.get('core.ignoredNames')
       follow: atom.config.get('core.followSymlinks')
 
-    task = Task.once require.resolve('./scan-handler'), atom.project.getPaths(), regex.source, searchOptions, ->
-      deferred.resolve()
+    # Find a search provider for every Directory in the project.
+    providersAndDirectories = []
+    for directory in atom.project.getDirectories()
+      providerForDirectory = null
+      for provider in @searchProviders
+        if provider.canSearchDirectory(directory)
+          providerForDirectory = provider
+          break
+      if providerForDirectory
+        providersAndDirectories.push({provider, directory})
+      else
+        throw Error("Could not find search provider for #{directory.getPath()}")
 
-    task.on 'scan:result-found', (result) ->
+    # Now that we are sure every Directory has a provider, construct the search options.
+    onSearchResult = (result) ->
       iterator(result) unless atom.project.isPathModified(result.filePath)
-
-    task.on 'scan:file-error', (error) ->
+    onSearchError = (error) ->
       iterator(null, error)
 
+    # Define the onPathsSearched callback.
     if _.isFunction(options.onPathsSearched)
-      task.on 'scan:paths-searched', (numberOfPathsSearched) ->
-        options.onPathsSearched(numberOfPathsSearched)
+      # Maintain a map of providers to the number of search results. When notified of a new count,
+      # replace the entry in the map and update the total.
+      onPathsSearchedOption = options.onPathsSearched
+      totalNumberOfPathsSearched = 0
+      numberOfPathsSearchedForProvider = new Map()
+      onPathsSearched = (provider, numberOfPathsSearched) ->
+        oldValue = numberOfPathsSearchedForProvider.get(provider)
+        if oldValue
+          totalNumberOfPathsSearched -= oldValue
+        numberOfPathsSearchedForProvider.set(provider, numberOfPathsSearched)
+        totalNumberOfPathsSearched += numberOfPathsSearched
+        onPathsSearchedOption(totalNumberOfPathsSearched)
+    else
+      onPathsSearched = ->
+
+    # Kick off all of the searches and unify them into one Promise.
+    allSearchPromises = []
+    for entry in providersAndDirectories
+      {provider, directory} = entry
+      recordNumPathsSearched = onPathsSearched.bind(undefined, provider)
```

Comment:
Seems a bit inconsistent to have `totalNumberOfPathsSearched` above but then shorten this to `recordNumPathsSearched` instead of `recordNumberOfPathsSearched`